### PR TITLE
Fixes #3589: Interface VLAN filtering

### LINKED ISSUE #3589: Missing error message when Tagged VLAN with incompatible Site is set
### Environment
* Python version: Python version: 3.6.8
* NetBox version: NetBox version: 2.6.5

### Steps to Reproduce
1. Create VLAN 123 associated with Site AAA
2. Create Device at Site BBB and Interface
3. Edit interface and set 802.1Q Mode to Tagged
4. Select VLAN 123 as Tagged VLAN
5. Apply

### Expected Behavior
An Error message.

### Observed Behavior
No error message, and the VLAN does not get set.

### Comment
A error message is correctly produced if the Mode is Access and the VLAN is selected in the Untagged input field instead. Then we get _"The untagged VLAN (123) must belong to the same site as the interface's parent device/VM, or it must be global"_. The same message should be show there too.

### PR DESCRIPTION
**Pull request by hSaria**
### Fixes: #3589

* Perform validation on the tagged VLANs similar to that which is already performed on the untagged VLAN
* Removes the legacy choices filter that is no longer used because the API-based select widgets override the choices
* Limit the select filter to only global VLANs or VLANs in the same site as the device

> Because there currently isn't a way to specify multiple values for the same parameter, I had to work around it.

### CODE REVIEW

File: netbox/dcim/forms.py [hunk 1]
```diff
@@ -74,6 +74,17 @@ def clean(self):
         elif self.cleaned_data['mode'] == IFACE_MODE_TAGGED_ALL:
             self.cleaned_data['tagged_vlans'] = []
 
+        # Validate tagged VLANs; must be a global VLAN or in the same site
+        else:
```
Comment (jeremystretch) [inline netbox/dcim/forms.py]:
```suggestion
        elif self.cleaned_data['mode'] == IFACE_MODE_TAGGED:
```
No need to check for this on access ports with no tagged VLANs.

File: netbox/dcim/forms.py [hunk 2]
```diff
@@ -2227,35 +2244,9 @@ def __init__(self, *args, **kwargs):
...
-                site_group_vlans = VLAN.objects.filter(group=group)
-                vlan_choices.append((
-                    '{} / {}'.format(group.site.name, group.name),
-                    [(vlan.pk, vlan) for vlan in site_group_vlans]
-                ))
-
-        self.fields['untagged_vlan'].choices = [(None, '---------')] + vlan_choices
-        self.fields['tagged_vlans'].choices = vlan_choices
+        # Add the current site to the list of filtered VLANs
+        self.fields['untagged_vlan'].widget.attrs['1-data-additional-query-param-site_id'] = self.instance.device.site.pk
```
Comment (jeremystretch) [inline netbox/dcim/forms.py]:
```suggestion
        self.fields['untagged_vlan'].widget.add_additional_query_param('site_id', self.instance.device.site.pk)
```
The widget has convenience methods for more cleanly applying these attributes.

File: netbox/dcim/forms.py [hunk 3]
```diff
@@ -2227,35 +2244,9 @@ def __init__(self, *args, **kwargs):
...
+        self.fields['tagged_vlans'].widget.attrs['1-data-additional-query-param-site_id'] = self.instance.device.site.pk
```
Comment (jeremystretch) [inline netbox/dcim/forms.py]:
```suggestion
        self.fields['tagged_vlans'].widget.add_additional_query_param('site_id', self.instance.device.site.pk)
```

File: netbox/dcim/forms.py [hunk 4]
```diff
@@ -2337,35 +2334,9 @@ def __init__(self, *args, **kwargs):
...
-                site_group_vlans = VLAN.objects.filter(group=group)
-                vlan_choices.append((
-                    '{} / {}'.format(group.site.name, group.name),
-                    [(vlan.pk, vlan) for vlan in site_group_vlans]
-                ))
-
-        self.fields['untagged_vlan'].choices = [(None, '---------')] + vlan_choices
-        self.fields['tagged_vlans'].choices = vlan_choices
+        # Add the current site to the list of filtered VLANs
+        self.fields['untagged_vlan'].widget.attrs['1-data-additional-query-param-site_id'] = self.parent.site.pk
```
Comment (jeremystretch) [inline netbox/dcim/forms.py]:
```suggestion
        self.fields['untagged_vlan'].widget.add_additional_query_param('site_id', self.parent.site.pk)
```

File: netbox/dcim/forms.py [hunk 5]
```diff
@@ -2337,35 +2334,9 @@ def __init__(self, *args, **kwargs):
...
+        self.fields['tagged_vlans'].widget.attrs['1-data-additional-query-param-site_id'] = self.parent.site.pk
```
Comment (jeremystretch) [inline netbox/dcim/forms.py]:
```suggestion
        self.fields['tagged_vlans'].widget.add_additional_query_param('site_id', self.parent.site.pk)
```

File: netbox/dcim/forms.py [hunk 6]
```diff
@@ -2449,35 +2426,9 @@ def __init__(self, *args, **kwargs):
...
-                    site_group_vlans = VLAN.objects.filter(group=group)
-                    vlan_choices.append((
-                        '{} / {}'.format(group.site.name, group.name),
-                        [(vlan.pk, vlan) for vlan in site_group_vlans]
-                    ))
-
-        self.fields['untagged_vlan'].choices = [(None, '---------')] + vlan_choices
-        self.fields['tagged_vlans'].choices = vlan_choices
+        # Add the current site to the list of filtered VLANs
+        self.fields['untagged_vlan'].widget.attrs['1-data-additional-query-param-site_id'] = self.parent_obj.site.pk
```
Comment (jeremystretch) [inline netbox/dcim/forms.py]:
```suggestion
        self.fields['untagged_vlan'].widget.add_additional_query_param('site_id', self.parent_obj.site.pk)
```

File: netbox/dcim/forms.py [hunk 7]
```diff
@@ -2449,35 +2426,9 @@ def __init__(self, *args, **kwargs):
...
+        self.fields['tagged_vlans'].widget.attrs['1-data-additional-query-param-site_id'] = self.parent_obj.site.pk
```
Comment (jeremystretch) [inline netbox/dcim/forms.py]:
```suggestion
        self.fields['tagged_vlans'].widget.add_additional_query_param('site_id', self.parent_obj.site.pk)
```

**Review by jeremystretch — CHANGES_REQUESTED** [review]

File: netbox/dcim/forms.py (hunk 3 above)
Comment (hSaria) [reply netbox/dcim/forms.py]: This wouldn't work because it will override

```
additional_query_params={
    'site_id': 'null'
},
```

**Comment by hSaria** [comment]
@jeremystretch FYI, this is currently waiting on your input regarding multiple values for the same field. The current implementation will override the field I've put in the `additional_query_params` of the widget.

I'm hoping there's a better way, but I can think of only two ways to do it:
* Key-based: use a different key which will map to the same field once the code in `forms.js` processes it (what I did, but perhaps in a better way than just prepending `1-`)
* Value-based: Use some character to indicate multiple values in the field (e.g. `add_additional_query_param('site_id', 'null,{}'.format(self.parent_obj.site.pk))`) and have `forms.js` split on that character.

**Comment by hSaria** [comment]
I have removed the select widget filtering changes from this PR as they will be covered in a different PR for #3840.

This PR just adds validation for the tagged VLANs of an interface.

**Review by jeremystretch — APPROVED** [review]

**Pull request merged**